src/main/java/com/exactpro/th2/common/schema/message/impl/rabbitmq/connection/ConnectionManager.java

```diff
@@ -732,30 +735,30 @@ public void withLock(boolean waitForRecovery, ChannelConsumer consumer) throws I
         }
 
         public void retryingPublishWithLock(ChannelConsumer consumer, ConnectionManagerConfiguration configuration) throws InterruptedException {
-            lock.lock();
             Iterator<RetryingDelay> iterator = configuration.createRetryingDelaySequence().iterator();
-            try {
-                Channel tempChannel = getChannel(true);
-                while (true) {
-                    try {
-                        consumer.consume(tempChannel);
-                        break;
-                    } catch (IOException | ShutdownSignalException e) {
-                        var currentValue = iterator.next();
-                        int recoveryDelay = currentValue.getDelay();
-                        LOGGER.warn("Retrying publishing #{}, waiting for {}ms. Reason: {}", currentValue.getTryNumber(), recoveryDelay, e);
-                        TimeUnit.MILLISECONDS.sleep(recoveryDelay);
-
-                        // We should not recover the channel if its connection is closed
-                        // If we do that the channel will be also auto recovered by RabbitMQ client
-                        // during connection recovery, and we will get two new channels instead of one closed.
-                        if (!tempChannel.isOpen() && tempChannel.getConnection().isOpen()) {
-                            tempChannel = recreateChannel();
-                        }
+            int recoveryDelay;
+            Channel tempChannel = getChannel(true);
+            while (true) {
+                lock.lock();
+                try {
+                    consumer.consume(tempChannel);
+                    break;
+                } catch (IOException | ShutdownSignalException e) {
+                    var currentValue = iterator.next();
+                    recoveryDelay = currentValue.getDelay();
+                    LOGGER.warn("Retrying publishing #{}, waiting for {}ms. Reason: {}", currentValue.getTryNumber(), recoveryDelay, e);
+
+                    // We should not recover the channel if its connection is closed
+                    // If we do that the channel will be also auto recovered by RabbitMQ client
+                    // during connection recovery, and we will get two new channels instead of one closed.
+                    if (!tempChannel.isOpen() && tempChannel.getConnection().isOpen()) {
+                        tempChannel = recreateChannel();
                     }
+                } finally {
+                    lock.unlock();
                 }
-            } finally {
-                lock.unlock();
+
+                TimeUnit.MILLISECONDS.sleep(recoveryDelay);
```

Comment:
Why this instruction was moved from catch block?

Reply:
To avoid locking for a long time.

Reply:
all operation with a channel are executed under lock. If one operation faces to a retry problem another will wait too.
In this case we can block another operations until resolve the retry problem

Reply:
ok, i will add one more lock for processing callbacks.

Reply:
No, please satay as is

Reply:
Several publishers follow replay schedule in your case.

Reply:
we need 2nd lock to read `channel` and `isSubscribed` values to filter out unnecessary callbacks without waiting for channel recovery. Otherwise all callbacks will be accumulated in executor service queue until the connection and channels will be recovered (and every recovery attempt generates new callback).

Reply:
We can use `volatile` or AtomicBoolean instead. Less lock, simple code, less chance to make a deadlock in future changes